File: 2018/05.py

import math, copy, re, hashlib
import itertools as it


def part_1(data):

	i = 0
	maxi = len(data) - 1
	while i < maxi:
		# print(i)
		if (data[i].isupper() == data[i + 1].islower() or data[i].islower() == data[i + 1].isupper()) and data[i].lower() == data[i + 1].lower():
			data = data[0:i] + data[i + 2:]
			maxi -= 2
			i -= 1
			if i < 0:
				i = 0
		else:
			i += 1

	print(len(data))
	print('END OF PART1')
	return

def part_2(data):

	shortest = len(data)

	for letter in range(ord('a'), ord('z') + 1):
		new_data = data.replace(chr(letter), '').replace(chr(letter).upper(), '')
		i = 0
		maxi = len(new_data) - 1
		while i < maxi:
			# print(i)
			if (new_data[i].isupper() == new_data[i + 1].islower() or new_data[i].islower() == new_data[i + 1].isupper()) and new_data[i].lower() == new_data[i + 1].lower():
				new_data = new_data[0:i] + new_data[i + 2:]
				maxi -= 2
				i -= 1
				if i < 0:
					i = 0
			else:
				i += 1		
		if (s := len(new_data)) < shortest:
			shortest = s
	print(shortest)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('05_input') as f:
		data = f.read()
		# data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2018/05.js

var fs = require('fs');
const { ord, chr, debug, disp, int, float, randint, gcd, lcm, crt, modPow, mod } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { PriorityQueue, Queue, set, Stack } = require("../lib");
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;

const sort = (arr, fn = (a, b) => a-b) => {
	arr.sort(fn);
}
const in_ = (a, arr) => arr.indexOf(a) !== -1;

const D4 = [[0,1],[1,0],[0,-1],[-1,0]];
const D8 = [[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1],[-1,0],[-1,1]];
const MOD = 1e9+7;

function part1(data) {
	let changes = true;
	while (changes) {
		changes = false;
		let i = 0;
		while (i < data.length - 1) {
			if (abs(ord(data[i]) - ord(data[i+1])) === abs(ord('A') - ord('a'))) {
				data = data.slice(0, i) + data.slice(i + 2);
				changes = true;
				continue;
			}
			++i;
		}
	}

	debug(data.length)

	console.log("END OF PART1");
	return;
}

function part2(_data) {
	let mn = _data.length;

	range(26)(i => {
		data = _data
			.replaceAll(chr(ord('a') + i), '')
			.replaceAll(chr(ord('A') + i), '');
		let changes = true;
		while (changes) {
			changes = false;
			let i = 0;
			while (i < data.length - 1) {
				if (abs(ord(data[i]) - ord(data[i+1])) === abs(ord('A') - ord('a'))) {
					data = data.slice(0, i) + data.slice(i + 2);
					changes = true;
					continue;
				}
				++i;
			}
		}
		if (data.length < mn) {
			mn = data.length;
		}
	})

	debug(mn)

	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("05_input").toString("utf-8");
	// data = data.split('\n');
	// data = data.split('\n').map(a => Number(a));

	if (Array.isArray(data)) {
		part1(Array.from(data));
		part2(Array.from(data));
	} else {
		part1(data);
		part2(data);
	}
}

main();
